File: wordSearch.py
def exist(board, word):
    rows, cols = len(board), len(board[0])
    
    def dfs(r, c, index):
        if index == len(word):  # If we matched all letters
            return True
        if r < 0 or r >= rows or c < 0 or c >= cols or board[r][c] != word[index]:
            return False
        
        # Mark the cell as visited by replacing it temporarily
        temp = board[r][c]
        board[r][c] = "#"  # Mark as visited
        
        # Explore in four directions: right, left, down, up
        found = (dfs(r+1, c, index+1) or 
                 dfs(r-1, c, index+1) or 
                 dfs(r, c+1, index+1) or 
                 dfs(r, c-1, index+1))
        
        # Restore the cell
        board[r][c] = temp
        return found
    
    for i in range(rows):
        for j in range(cols):
            if board[i][j] == word[0] and dfs(i, j, 0):  # Start DFS if first letter matches
                return True
    
    return False

# Example Usage
board = [["A", "B", "C", "E"],
         ["S", "F", "C", "S"],
         ["A", "D", "E", "E"]]

word = "ABCCED"
print(exist(board, word))  # Output: True

File: wordSearch.cpp
#include<iostream>
#include<vector>
using namespace std;

bool dfs(vector<vector<char>> board, string word, int i,int j, int index){
    if(index==word.length()) return true;
    if(i<0 || i>=board.size() || j<0 || j>=board.size() || board[i][j] != word[index]){
        return false;
    }
    
    char temp = board[i][j];
    board[i][j] = '#';

    bool found = dfs(board, word, i+1, j, index+1)||
                 dfs(board, word, i-1, j, index+1) ||
                 dfs(board, word, i, j+1, index+1) ||
                 dfs(board, word, i,j-1, index+1);
    
    board[i][j] = temp;
    return found;
}

bool exist(vector<vector<char>> board, string word)
{
    int index = 0;
    for(int i=0; i<board.size(); i++){
        for(int j=0; j<board[i].size(); j++){
            if(dfs(board, word, i, j, 0)){
                return true;
                } 
            }
        }
        return false;
}


int main(){
    vector<vector<char>> board = {{"A", "B", "C", "E"},
                                 {"S", "F", "C", "S"},
                                 {"A", "D", "E", "E"}};
    string word = "ABCCED"; 
    cout<<(exist(board, word) ? "True" : "False")<<endl;
    return 0;

}


/*
1. find the first letter
2. check for all adjacent blocks(left, right up, down)
3. if next letter found, letter is the next letter
4. continue until letter is the last letter
*/
